### PR TITLE
RavenDB-18043 stream results from shards

### PR DESCRIPTION
**Pull request by karmeli87**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18043

### Additional description

Add the functionality to stream the results from the shards

### Type of change

- Optimization

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Client/Documents/Session/AsyncDocumentSession.TimeSeriesStream.cs [hunk 1]
```diff
@@ -43,10 +44,10 @@ internal class YieldTimeSeriesStream<T> : AbstractYieldStream<TimeSeriesStreamRe
                 _query = query;
             }
 
-            internal override TimeSeriesStreamResult<T> ResultCreator(StreamOperation.YieldStreamResults asyncEnumerator)
+            internal override TimeSeriesStreamResult<T> ResultCreator(IAsyncEnumerator<BlittableJsonReaderObject> asyncEnumerator)
             {
                 _query?.InvokeAfterStreamExecuted(asyncEnumerator.Current);
-                return _parent.CreateTimeSeriesStreamResult<T>(asyncEnumerator);
+                return _parent.CreateTimeSeriesStreamResult<T>((StreamOperation.YieldStreamResults)asyncEnumerator);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/AsyncDocumentSession.TimeSeriesStream.cs]: can we fix that as well?

File: src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewCollection.cs [hunk 2]
```diff
@@ -43,37 +44,42 @@ protected AbstractStudioCollectionsHandlerProcessorForPreviewCollection([NotNull
         HttpContext = requestHandler.HttpContext;
     }
 
-    protected virtual void Initialize()
+    protected virtual ValueTask Initialize()
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewCollection.cs]: Async

File: src/Raven.Server/Web/Studio/Processors/StudioCollectionsHandlerProcessorForPreviewCollection.cs [hunk 3]
```diff
@@ -111,7 +113,7 @@ public override void Dispose()
         _releaseContext = null;
     }
 
-    private static List<string> ExtractColumnNames(List<Document> documents, JsonOperationContext context)
+    public static List<string> ExtractColumnNames(List<Document> documents, JsonOperationContext context)
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Processors/StudioCollectionsHandlerProcessorForPreviewCollection.cs]: why?

File: src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs [hunk 4]
```diff
@@ -60,135 +91,89 @@ protected override bool NotModified(out string etag)
...
     {
+        _combinedReadState.Dispose();
         base.Dispose();
 
         _releaseContext?.Dispose();
         _releaseContext = null;
     }
 
-    private class PreviewCollectionResult
+    public class PreviewCollectionResult
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs]: why?

File: src/Raven.Server/Web/Studio/Sharding/ShardedStudioCollectionsHandler.cs [hunk 5]
```diff
@@ -1,9 +1,9 @@
 ﻿using System.Threading.Tasks;
-using Raven.Server.Documents.Sharding;
+ using Raven.Server.Documents.Sharding;
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Sharding/ShardedStudioCollectionsHandler.cs]: hm?

File: src/Raven.Server/Web/Studio/Sharding/ShardedStudioCollectionsHandler.cs [hunk 6]
```diff
@@ -1,9 +1,9 @@
...
 using Raven.Server.Routing;
 using Raven.Server.Web.Studio.Sharding.Processors;
 
-namespace Raven.Server.Web.Studio.Sharding
+ namespace Raven.Server.Web.Studio.Sharding
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Sharding/ShardedStudioCollectionsHandler.cs]: hm?

File: src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs [hunk 7]
```diff
@@ -60,135 +91,89 @@ protected override bool NotModified(out string etag)
...
-                var shardDocs = span[takeFromShard].Results;
-                if (shardDocs.Count == 0)
-                    continue;
-
-                if (totalDocuments++ >= _token.PageSize)
-                    return total;
+                Result = new StreamResult
+                {
+                    Response = response,
+                    Stream = new StreamWithTimeout(responseStream)
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs:152]: do we want to wrap it? there is a read timeout there and I'm assuming that some of the streams might not be read for a long time?

File: src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs [hunk 8]
```diff
@@ -60,135 +91,89 @@ protected override bool NotModified(out string etag)
...
+        {
+            results.Add(doc.Item.Clone(_context));
+        }
 
-        return _result.Results;
+        return results;
     }
 
-    protected override List<string> GetAvailableColumns(List<Document> documents)
+    protected override async ValueTask<List<string>> GetAvailableColumns(List<Document> documents)
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs]: Async

File: src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs [hunk 9]
```diff
@@ -60,135 +91,89 @@ protected override bool NotModified(out string etag)
...
+        {
+            total.AddRange(result.Span[i].Item);
+        }
+
+        return total.ToList();
     }
 
     public override void Dispose()
     {
+        _combinedReadState.Dispose();
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs]: null check + set it to null?

File: src/Raven.Server/Web/Studio/Processors/StudioCollectionsHandlerProcessorForPreviewCollection.cs [hunk 10]
```diff
@@ -95,9 +97,9 @@ protected override ValueTask<List<Document>> GetDocumentsAsync()
         return ValueTask.FromResult(documents);
     }
 
-    protected override List<string> GetAvailableColumns(List<Document> documents)
+    protected override ValueTask<List<string>> GetAvailableColumns(List<Document> documents)
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Processors/StudioCollectionsHandlerProcessorForPreviewCollection.cs]: Async

File: src/Raven.Server/Web/Studio/Processors/StudioCollectionsHandlerProcessorForPreviewCollection.cs [hunk 11]
```diff
@@ -31,7 +31,7 @@ public StudioCollectionsHandlerProcessorForPreviewCollection(DatabaseRequestHand
         _database = database ?? throw new ArgumentNullException(nameof(database));
     }
 
-    protected override void Initialize()
+    protected override ValueTask Initialize()
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Processors/StudioCollectionsHandlerProcessorForPreviewCollection.cs]: Async

File: src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewCollection.cs [hunk 12]
```diff
@@ -43,37 +44,42 @@ protected AbstractStudioCollectionsHandlerProcessorForPreviewCollection([NotNull
...
+    protected abstract ValueTask<List<string>> GetAvailableColumns(List<Document> documents);
 
     public async Task ExecuteAsync()
     {
-        Initialize();
+        await Initialize();
 
         if (NotModified(out var etag))
+        {
+            RequestHandler.HttpContext.Response.StatusCode = (int)HttpStatusCode.NotModified;
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewCollection.cs:73]: thx

File: src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewCollection.cs [hunk 13]
```diff
@@ -86,36 +92,47 @@ public async Task ExecuteAsync()
...
         {
             writer.WriteStartObject();
-            writer.WritePropertyName("Results");
+            WriteResult(writer, documents, context, propertiesPreviewToSend, fullPropertiesToSend, totalResults, availableColumns);
+            writer.WriteEndObject();
+        }
+    }
 
-            writer.WriteStartArray();
+    protected virtual void WriteResult(
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewCollection.cs]: WriteResultsAsync

Comment (ppekrol) [reply src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewCollection.cs]: WriteResults, sorry*

File: src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs [hunk 14]
```diff
@@ -60,135 +91,89 @@ protected override bool NotModified(out string etag)
 
     protected override async ValueTask<List<Document>> GetDocumentsAsync()
     {
-        var op = new ShardedCollectionPreviewOperation(RequestHandler, _continuationToken, _context);
-        _result = await RequestHandler.ShardExecutor.ExecuteParallelForAllAsync(op);
+        var results = new List<Document>();
+
+        await foreach (var doc in _requestHandler.ShardedContext.Streaming.PagedShardedDocumentsByLastModified(_combinedReadState, nameof(PreviewCollectionResult.Results), _continuationToken))
+        {
+            results.Add(doc.Item.Clone(_context));
```
Comment (ppekrol) [inline src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs]: why we need to clone here? is `doc` released after moving to next element?

File: src/Raven.Server/Documents/ShardedHandlers/ContinuationTokens/ShardedPagingContinuation.cs [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        return new DynamicJsonValue
+        {
+            [nameof(Pages)] = new DynamicJsonArray(Pages), 
+            [nameof(PageSize)] = PageSize
+        };
+    }
+
+    public struct ShardPaging : IDynamicJson
+    {
+        public int Shard;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ShardedHandlers/ContinuationTokens/ShardedPagingContinuation.cs]: ShardNumber

File: src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewCollection.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewCollection.cs]: done

File: src/Raven.Server/Web/Studio/Processors/StudioCollectionsHandlerProcessorForPreviewCollection.cs (hunk 3 above)
Comment (karmeli87) [reply src/Raven.Server/Web/Studio/Processors/StudioCollectionsHandlerProcessorForPreviewCollection.cs]: done

File: src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs (hunk 4 above)
Comment (karmeli87) [reply src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs]: done

File: src/Raven.Server/Web/Studio/Sharding/ShardedStudioCollectionsHandler.cs (hunk 5 above)
Comment (karmeli87) [reply src/Raven.Server/Web/Studio/Sharding/ShardedStudioCollectionsHandler.cs]: done

File: src/Raven.Server/Web/Studio/Sharding/ShardedStudioCollectionsHandler.cs (hunk 6 above)
Comment (karmeli87) [reply src/Raven.Server/Web/Studio/Sharding/ShardedStudioCollectionsHandler.cs]: done

File: src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs (hunk 8 above)
Comment (karmeli87) [reply src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs]: done

File: src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs (hunk 9 above)
Comment (karmeli87) [reply src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs]: done

File: src/Raven.Server/Web/Studio/Processors/StudioCollectionsHandlerProcessorForPreviewCollection.cs (hunk 10 above)
Comment (karmeli87) [reply src/Raven.Server/Web/Studio/Processors/StudioCollectionsHandlerProcessorForPreviewCollection.cs]: done

File: src/Raven.Server/Web/Studio/Processors/StudioCollectionsHandlerProcessorForPreviewCollection.cs (hunk 11 above)
Comment (karmeli87) [reply src/Raven.Server/Web/Studio/Processors/StudioCollectionsHandlerProcessorForPreviewCollection.cs]: done

File: src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewCollection.cs (hunk 13 above)
Comment (karmeli87) [reply src/Raven.Server/Web/Studio/Processors/AbstractStudioCollectionsHandlerProcessorForPreviewCollection.cs]: done

File: src/Raven.Client/Documents/Session/AsyncDocumentSession.TimeSeriesStream.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Client/Documents/Session/AsyncDocumentSession.TimeSeriesStream.cs]: done

File: src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs (hunk 7 above)
Comment (karmeli87) [reply src/Raven.Server/Web/Studio/Sharding/Processors/ShardedStudioCollectionsHandlerProcessorForPreviewCollection.cs:152]: This is not something that we should do across the board, but I do expect the studio to show me the results after 2 min..

**Comment by ppekrol** [comment]
@ml054 please review Studio work and issue a separate PR if needed.

File: src/Raven.Studio/typescript/commands/database/documents/getDocumentsPreviewCommand.ts [hunk 16]
```diff
@@ -28,26 +28,38 @@ class getDocumentsPreviewCommand extends commandBase {
...
     }
 
     execute(): JQueryPromise<pagedResultWithAvailableColumns<document>> {
-        const args = {
+        
+        const argsWithContinuationToken = {
+            collection: this.collectionName,
+            binding: this.previewBindings,
+            fullBinding: this.fullBindings,
+            "continuation-token": this.continuationToken
```
Comment (ml054) [inline src/Raven.Studio/typescript/commands/database/documents/getDocumentsPreviewCommand.ts:42]: @karmeli87  Don't we use camelCase for url arguments?

File: src/Raven.Studio/typescript/commands/database/documents/getDocumentsPreviewCommand.ts [hunk 17]
```diff
@@ -28,26 +28,38 @@ class getDocumentsPreviewCommand extends commandBase {
...
-    constructor(private database: database, private skip: number, private take: number, private collectionName: string, private previewBindings?: string[], private fullBindings?: string[]) {
+    constructor(private database: database, private skip: number, private take: number, private collectionName: string,
+                private previewBindings?: string[], private fullBindings?: string[], private continuationToken?: string) {
         super();
     }
 
     execute(): JQueryPromise<pagedResultWithAvailableColumns<document>> {
-        const args = {
+        
+        const argsWithContinuationToken = {
```
Comment (ml054) [inline src/Raven.Studio/typescript/commands/database/documents/getDocumentsPreviewCommand.ts:38]: @Danielle9897 extract this to separate method. Right now you  create 2 objects and choose one.

File: src/Raven.Studio/typescript/viewmodels/database/documents/documents.ts [hunk 18]
```diff
@@ -198,9 +202,39 @@ class documents extends viewModelBase {
...
+                        results.totalResultCount = this.itemsSoFar()
+                    } else {
+                        results.totalResultCount = this.itemsSoFar() + 1;
+                    }
+                }
+            }));
     }
 
+    private getContinuationTokenToUse() {
+        const isShardedDatabase = shardedDatabase.isSharded(this.db);
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/documents/documents.ts:225]: I wouldn't use that - if server returns continution token then use it. W/o any extra logic.

File: src/Raven.Studio/typings/_studio/dto.d.ts [hunk 19]
```diff
@@ -57,6 +57,7 @@ interface resultsWithTotalCountDto<T> extends resultsDto<T> {
 
 interface resultsWithCountAndAvailableColumns<T> extends resultsWithTotalCountDto<T> {
     AvailableColumns: string[];
+    ContinuationToken: string;
```
Comment (ml054) [inline src/Raven.Studio/typings/_studio/dto.d.ts:60]: ContinuationToken?: string;

File: src/Raven.Studio/typings/_studio/dto.d.ts [hunk 20]
```diff
@@ -442,6 +443,7 @@ interface pagedResultExtended<T> extends pagedResult<T> {
 
 interface pagedResultWithAvailableColumns<T> extends pagedResult<T> {
     availableColumns: string[];
+    continuationToken: string;
```
Comment (ml054) [inline src/Raven.Studio/typings/_studio/dto.d.ts:446]: continuationToken?: string;

File: src/Raven.Studio/typescript/commands/database/documents/getDocumentsPreviewCommand.ts (hunk 16 above)
Comment (karmeli87) [reply src/Raven.Studio/typescript/commands/database/documents/getDocumentsPreviewCommand.ts:42]: @Danielle9897 as agreed with marcin please change this to camelCase and verify it is working as expected.

You should also modify on the server side this line:
`public const string ContinuationTokenQueryString = "continuation-token";`

File: src/Raven.Studio/typescript/commands/database/documents/getDocumentsPreviewCommand.ts (hunk 17 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/commands/database/documents/getDocumentsPreviewCommand.ts:38]: done

File: src/Raven.Studio/typings/_studio/dto.d.ts (hunk 20 above)
Comment (Danielle9897) [reply src/Raven.Studio/typings/_studio/dto.d.ts:446]: done

File: src/Raven.Studio/typings/_studio/dto.d.ts (hunk 19 above)
Comment (Danielle9897) [reply src/Raven.Studio/typings/_studio/dto.d.ts:60]: done

File: src/Raven.Studio/typescript/viewmodels/database/documents/documents.ts (hunk 18 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/database/documents/documents.ts:225]: done

File: src/Raven.Studio/typescript/commands/database/documents/getDocumentsPreviewCommand.ts (hunk 16 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/commands/database/documents/getDocumentsPreviewCommand.ts:42]: done

**Pull request merged**